File: app/src/main/java/org/mesonet/app/webview/WebViewActivity.java
package org.mesonet.app.webview;

import android.content.Intent;
import android.databinding.DataBindingUtil;
import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;
import android.view.LayoutInflater;
import android.view.View;
import android.webkit.WebSettings;

import org.mesonet.app.R;
import org.mesonet.app.databinding.WebViewActivityBinding;

import java.io.BufferedReader;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.net.HttpURLConnection;
import java.net.MalformedURLException;
import java.net.URL;


public class WebViewActivity extends AppCompatActivity
{
    public static final String kTitle = "title";
    public static final String kUrl = "url";
    public static final String kInitialZoom = "initialZoom";
    public static final String kUseGoogleDocs = "useGoogleDocs";
    public static final String kAllowShare = "allowShare";


    @Override
    public void onCreate(Bundle inSavedInstanceState)
    {
        super.onCreate(inSavedInstanceState);
        final WebViewActivityBinding binding = DataBindingUtil.inflate(LayoutInflater.from(this), R.layout.web_view_activity, null, false);

        final String title = getIntent().getStringExtra(kTitle);
        String url = getIntent().getStringExtra(kUrl);

        if(getIntent().getBooleanExtra(kUseGoogleDocs, false))
            url = "http://docs.google.com/gview?embedded=true&url=" + url;

        WebSettings webSettings = binding.webView.getSettings();
        webSettings.setJavaScriptEnabled(true);

        binding.toolBar.setTitle(title);
        binding.toolBar.setNavigationIcon(R.drawable.ic_close_white_36dp);
        binding.toolBar.setNavigationOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                finish();
            }
        });

        final String finalUrl = url;

        if(getIntent().getBooleanExtra(kAllowShare, false)) {
            binding.shareButton.setVisibility(View.VISIBLE);
            binding.shareButton.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    Intent i = new Intent(Intent.ACTION_SEND);
                    i.setType("text/plain");
                    i.putExtra(Intent.EXTRA_SUBJECT, title);
                    i.putExtra(Intent.EXTRA_TEXT, finalUrl);
                    startActivity(Intent.createChooser(i, "Share URL"));
                }
            });
        }

        final String webInfoUrl = url;

        Integer initialScale = getIntent().getIntExtra(kInitialZoom, -1);

        if(initialScale != -1) {
            binding.webView.setInitialScale(initialScale);

            binding.webView.getSettings().setLoadWithOverviewMode(true);
            binding.webView.getSettings().setUseWideViewPort(true);
        }

//        if(webInfoUrl.endsWith(".txt"))
//        {
//            new Thread(new Runnable() {
//                @Override
//                public void run() {
//                    try {
//
//                        String text = "";
//                        URL url = new URL(webInfoUrl);
//                        HttpURLConnection con=(HttpURLConnection)url.openConnection();
//                        //get InputStream instance
//                        InputStream is=con.getInputStream();
//                        //create BufferedReader object
//                        BufferedReader br=new BufferedReader(new InputStreamReader(is));
//                        String line;
//                        //read content of the file line by line
//                        while((line=br.readLine())!=null){
//                            text+=line;
//
//                        }
//
//                        br.close();
//
//                        final StringBuilder sb = new StringBuilder();
//                        String[] words = text.toString().split("\\s");  // assume no spaces in links
//
//                        for (String word : words) {
//                            try {
//                                // Attempt to add the string as a link
//                                url = new URL(word);
//                                sb.append("<p>");
//                                sb.append("<a href=\"");
//                                sb.append(url.toString());
//                                sb.append("\">");
//                                sb.append(url);
//                                sb.append("</a> ");
//                                sb.append("</p>");
//                            } catch (MalformedURLException e) {
//                                // This was not a valid URL, just add it to the string
//                                sb.append(word);
//                                sb.append(" ");
//                            }
//                        }
//
//                        runOnUiThread(new Runnable() {
//                            @Override
//                            public void run() {
//                                binding.webView.loadData(sb.toString(), "text/html", "UTF-8");
//                            }
//                        });
//                    }
//                    catch (Exception e)
//                    {
//                        runOnUiThread(new Runnable() {
//                            @Override
//                            public void run() {
//                                binding.webView.loadUrl(finalUrl);
//                            }
//                        });
//                    }
//                }
//            }).start();
//        }
//        else
            binding.webView.loadUrl(url);

        setContentView(binding.getRoot());
    }
}

File: app/src/main/java/org/mesonet/app/webview/WebViewActivity.kt
package org.mesonet.app.webview

import android.content.Intent
import android.databinding.DataBindingUtil
import android.os.Bundle
import android.os.Handler
import android.support.v4.view.PagerAdapter
import android.support.v4.view.ViewPager
import android.support.v7.app.AppCompatActivity
import android.view.LayoutInflater
import android.view.View
import android.webkit.WebViewClient
import org.mesonet.app.R

import org.mesonet.app.databinding.WebViewActivityBinding
import java.util.*
import android.view.ViewGroup
import android.webkit.WebView


class WebViewActivity : AppCompatActivity() {
    var mPaused = true
    var mLastUpdate = 0L
    var mBinding: WebViewActivityBinding? = null

    var mAdapter: WebViewPagerAdapter? = null

    public override fun onCreate(inSavedInstanceState: Bundle?) {
        super.onCreate(inSavedInstanceState)
        mBinding = DataBindingUtil.inflate(LayoutInflater.from(this), R.layout.web_view_activity, null, false)

        var urls: List<String>? = null
        var titles: List<String>? = null
        var startIndex = 0

        mAdapter = WebViewPagerAdapter()

        if(intent.hasExtra(kUrls))
            urls = intent.getStringArrayListExtra(kUrls)

        if(intent.hasExtra(kTitles))
            titles = intent.getStringArrayListExtra(kTitles)

        if(intent.hasExtra(kStartIndex))
            startIndex = intent.getIntExtra(kStartIndex, 0)

        val pagerList = ArrayList<PagerData>()

        mBinding?.toolBar?.setNavigationIcon(R.drawable.ic_close_white_36dp)
        mBinding?.toolBar?.setNavigationOnClickListener { finish() }

        for(i in 0 until (urls?.size?: 0)) {
            var url = urls?.get(i)
            if (intent.getBooleanExtra(kUseGoogleDocs, false))
                url = "http://docs.google.com/gview?embedded=true&url=$url"// + URLEncoder.encode(url, "UTF-8")

            pagerList.add(PagerData(titles?.get(i)?: "", url))
        }

        if (intent.getBooleanExtra(kAllowShare, false)) {
            mBinding?.shareButton?.show()
            mBinding?.shareButton?.setOnClickListener {
                it.isEnabled = false
                val i = Intent(Intent.ACTION_SEND)
                i.type = "text/plain"

                (mBinding?.viewPager?.adapter as WebViewPagerAdapter).mData?.let {data ->
                    if(data.isNotEmpty()) {
                        val selectedMap = data[mBinding?.viewPager?.currentItem ?: 0]
                        i.putExtra(Intent.EXTRA_SUBJECT, selectedMap.mTitle)
                        i.putExtra(Intent.EXTRA_TEXT, selectedMap.mUrl)
                    }
                }

                startActivity(Intent.createChooser(i, "Share URL"))
                it.isEnabled = true
            }
        }

        if (urls == null) {
            if (intent.hasExtra(kRaw) && !intent.getStringExtra(kRaw).isBlank()) {
                var title = ""
                if(titles?.size?: 0 > 0)
                    title = titles?.get(0)?: ""

                mAdapter?.SetItems(listOf(PagerData(title, null)))
            }
        }
        else
            mAdapter?.SetItems(pagerList)

        mLastUpdate = Date().time

        mBinding?.viewPager?.adapter = mAdapter

        val pageChangeListener = object: ViewPager.OnPageChangeListener {
            override fun onPageScrollStateChanged(p0: Int) {}
            override fun onPageScrolled(p0: Int, p1: Float, p2: Int) {}

            override fun onPageSelected(inPage: Int) {
                val title = (mBinding?.viewPager?.adapter as WebViewPagerAdapter).mData?.get(inPage)?.mTitle

                mBinding?.toolBar?.title = title
            }

        }

        mBinding?.viewPager?.addOnPageChangeListener(pageChangeListener)

        mBinding?.viewPager?.postDelayed({
                if(startIndex == 0)
                    pageChangeListener.onPageSelected(0)
                else
                    mBinding?.viewPager?.setCurrentItem(startIndex, false)


        }, 100)

        setContentView(mBinding?.root)
    }

    override fun onResume() {
        super.onResume()

        val updateInterval = intent.getLongExtra(kUpdateInterval, -1)

        if(mPaused) {
            mPaused = false
            if (updateInterval > -1) {
                Handler().post(object : Runnable {
                    override fun run() {
                        if (!mPaused) {
                            if(mLastUpdate - updateInterval < Date().time) {
                                (mBinding?.viewPager?.adapter as WebViewPagerAdapter).RefreshPages()
                                mLastUpdate = Date().time
                            }
                            Handler().postDelayed(this, updateInterval)
                        }
                    }
                })
            }
        }
    }

    override fun onPause() {
        mPaused = true

        super.onPause()
    }

    companion object {
        val kTitles = "titles"
        val kUrls = "urls"
        val kStartIndex = "startIndex"
        val kRaw = "raw"
        val kInitialZoom = "initialZoom"
        val kAllowUserZoom = "allowUserZoom"
        val kUseGoogleDocs = "useGoogleDocs"
        val kAllowShare = "allowShare"
        val kUpdateInterval = "updateInterval"
    }


    inner class WebViewPagerAdapter: PagerAdapter() {
        var mData: List<PagerData>? = null
        var mWebViews: ArrayList<WebView>? = null

        fun SetItems(pagerData: List<PagerData>) {
            mData = pagerData
            mWebViews = arrayListOf()
            notifyDataSetChanged()
        }

        override fun instantiateItem(collection: ViewGroup, position: Int): Any {
            val layout = layoutInflater.inflate(R.layout.web_view, collection, false)

            val webView = layout.findViewById<WebView>(R.id.webView)

            val webSettings = webView.settings
            webSettings?.javaScriptEnabled = true
            webView.webViewClient = WebViewClient()

            if (intent.getBooleanExtra(kAllowUserZoom, false)) {
                webView.settings?.setSupportZoom(true)
                webView.settings?.builtInZoomControls = true
            }

            val initialScale = intent.getIntExtra(kInitialZoom, -1)

            if (initialScale != -1) {
                webView.setInitialScale(initialScale)

                webView.settings?.loadWithOverviewMode = true
                webView.settings?.useWideViewPort = true
            }

            collection.addView(layout)

            mWebViews?.add(webView)

            mData?.get(position)?.mUrl?.let {
                webView.loadUrl(it)
                return layout
            }

            if(intent.hasExtra(kRaw))
                webView.loadUrl(intent.getStringExtra(kRaw))

            return layout
        }

        fun RefreshPages() {
            mWebViews?.forEach { webView ->
                webView.reload()
            }
        }

        override fun destroyItem(inContainer: ViewGroup, inPosition: Int, inObject: Any) {
            inContainer.removeView(inObject as View)
        }

        override fun isViewFromObject(inView: View, inObject: Any): Boolean {
            return inView == inObject
        }

        override fun getCount(): Int {
            return mData?.size?: 0
        }
    }

    class PagerData (
        val mTitle: String,
        val mUrl: String?
    )
}
